src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpRequestMessageExtensions.cs

```diff
@@ -39,38 +38,48 @@ public static HttpRequestMessage Clone(this HttpRequestMessage request)
         }
 
         /// <summary>
-        /// Retrieves a logger instance attached to the given request as custom property.
+        /// Retrieves the HTTP request configuration instance attached to the given message as custom property.
         /// </summary>
-        /// <param name="request">Request message</param>
-        /// <returns>Logger instance if exists, or null otherwise.</returns>
-        public static ILogger GetLogger(this HttpRequestMessage request)
+        /// <param name="request">The HTTP request message.</param>
+        /// <returns>Configuration instance if exists, or a default instance otherwise.</returns>
+        public static HttpRequestMessageConfiguration GetConfiguration(this HttpRequestMessage request)
         {
             if (request == null)
             {
                 throw new ArgumentNullException(nameof(request));
             }
 
-            return request.GetProperty<ILogger>(NuGetLoggerKey);
+            var foundInstance = request.GetProperty<HttpRequestMessageConfiguration>(NuGetConfigurationKey);
+
+            if (foundInstance == null)
+            {
+                foundInstance = HttpRequestMessageConfiguration.Default;
+                request.SetConfiguration(foundInstance);
```

Comment:
Having lazily initialized properties is a pattern all over NuGet code. I don't see a big problem here (or even a good alternative without propagating null handling to all callers which is worse in my mind), especially considering the fact that we don't control the ctor in this case.